### PR TITLE
The big refactor

### PR DESCRIPTION
**Pull request by George3d6**

### CODE REVIEW

File: mindsdb/interfaces/model/learn_process.py [hunk 1]
```diff
@@ -29,77 +31,44 @@ def delete_learn_mark():
...
-
-    to_predict = to_predict if isinstance(to_predict, list) else [to_predict]
-    data_source = getattr(mindsdb_datasources, from_data['class'])(*from_data['args'], **from_data['kwargs'])
     try:
-        mdb.learn(
-            from_data=data_source,
-            to_predict=to_predict,
-            **kwargs
-        )
+        predictor_record = session.query(db.Predictor).filter_by(id=preidctor_id).first()
```
Comment (StpMax) [inline mindsdb/interfaces/model/learn_process.py:38]: ```session.query(db.Predictor).get(preidctor_id)``` looks a little better

File: mindsdb/interfaces/model/learn_process.py [hunk 2]
```diff
@@ -29,77 +31,44 @@ def delete_learn_mark():
...
-        delete_process_mark('learn')
+        fs_store = FsSotre()
+        config = Config()
+        
+        predictor: lightwood.PredictorInterface = lightwood.predictor_from_code(predictor_record.code)
+        predictor.learn(df)
 
-    if save:
-        fs_store.put(name, f'predictor_{company_id}_{predictor_record.id}', config['paths']['predictors'])
+        predictor_record = session.query(db.Predictor).filter_by(id=preidctor_id).first()
```
Comment (StpMax) [inline mindsdb/interfaces/model/learn_process.py:47]: ```session.refresh(instance_record)``` looks a little better

File: mindsdb/interfaces/model/model_controller.py [hunk 3]
```diff
@@ -1,366 +1,383 @@
...
-                                         MySqlDS, PostgresDS, MSSQLDS, MongoDS,
-                                         SnowflakeDS, AthenaDS)
-        import mindsdb_native
-        from mindsdb.interfaces.storage.db import session, Predictor
+        self.generate_predictor(name, from_data, datasource_id, problem_definition, company_id)
 
+        # TODO: Should we support kwargs['join_learn_process'](?)
+        self.fit_predictor(name, from_data, join_learn_process, company_id)
+
+    def predict(self, name: str, when_data: dict, pred_format: str, company_id: int):
```
Comment (StpMax) [inline mindsdb/interfaces/model/model_controller.py:98]: ```when_data: Union[dict, list]``` ?

File: mindsdb/interfaces/model/model_controller.py [hunk 4]
```diff
@@ -1,366 +1,383 @@
...
+
+        predictor_record = db.session.query(db.Predictor).filter_by(company_id=company_id, name=name).first()
+        assert predictor_record is not None
+        predictor_record.data = {'status': 'training', 'name': name}
+        db.session.commit()
+
+        p = LearnProcess(predictor_record.id, df)
+        p.start()
+        if join_learn_process:
+            p.join()
```
Comment (StpMax) [inline mindsdb/interfaces/model/model_controller.py:380]: need add `delete_process_mark('learn')` to the end

File: mindsdb/interfaces/model/learn_process.py (hunk 2 above)
Comment (George3d6) [reply mindsdb/interfaces/model/learn_process.py:47]: Oh, forgot that was a thing, cheers :)

File: mindsdb/interfaces/model/model_controller.py (hunk 3 above)
Comment (George3d6) [reply mindsdb/interfaces/model/model_controller.py:98]: Should be `Union[dict, list, pd.DataFrame]` even, well spotted

File: mindsdb/interfaces/model/model_controller.py (hunk 4 above)
Comment (George3d6) [reply mindsdb/interfaces/model/model_controller.py:380]: Should already be happening in the learn process so not needed

**Pull request merged**